File: app/src/main/java/com/example/waveplayer/fragments/FragmentSelectSongs.java
package com.example.waveplayer.fragments;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.waveplayer.activity_main.ActivityMain;
import com.example.waveplayer.activity_main.ViewModelActivityMain;
import com.example.waveplayer.databinding.RecyclerViewSongListBinding;
import com.example.waveplayer.random_playlist.Song;
import com.example.waveplayer.ViewModelUserPickedSongs;
import com.example.waveplayer.R;

import java.util.ArrayList;
import java.util.List;

public class FragmentSelectSongs extends Fragment
        implements RecyclerViewAdapterSelectSongs.ListenerCallbackSelectSongs {

    public static final String NAME = "FragmentSelectSongs";

    private RecyclerViewSongListBinding binding;

    private ViewModelActivityMain viewModelActivityMain;
    private ViewModelUserPickedSongs viewModelUserPickedSongs;

    private BroadcastReceiver broadcastReceiver;

    private RecyclerView recyclerViewSongList;
    private RecyclerViewAdapterSelectSongs recyclerViewAdapter;

    private View.OnClickListener onClickListenerFABFragmentSelectSongs;
    private SearchView.OnQueryTextListener onQueryTextListenerSearch;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        createViewModels();
        binding = RecyclerViewSongListBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    private void createViewModels() {
        viewModelUserPickedSongs =
                new ViewModelProvider(requireActivity()).get(ViewModelUserPickedSongs.class);
        viewModelActivityMain =
                new ViewModelProvider(requireActivity()).get(ViewModelActivityMain.class);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ActivityMain activityMain = (ActivityMain) requireActivity();
        activityMain.hideKeyboard(view);
        viewModelActivityMain.setActionBarTitle(getResources().getString(R.string.select_songs));
        setUpBroadcastReceiver();
        setUpRecyclerView();
    }

    private void setUpBroadcastReceiver() {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT);
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_service_connected));
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_on_create_options_menu));
        broadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();
                if (action != null) {
                    if (action.equals(getResources().getString(
                            R.string.broadcast_receiver_action_on_create_options_menu))) {
                        setUpToolbar();
                    } else if (action.equals(getResources().getString(
                            R.string.broadcast_receiver_action_service_connected))) {
                        setUpRecyclerView();
                    }
                }
            }
        };
        activityMain.registerReceiver(broadcastReceiver, intentFilter);
    }

    private void setUpToolbar() {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        Toolbar toolbar = activityMain.findViewById(R.id.toolbar);
        Menu menu = toolbar.getMenu();
        if (menu != null) {
            menu.getItem(ActivityMain.MENU_ACTION_SEARCH_INDEX).setVisible(true);
            MenuItem itemSearch = menu.findItem(R.id.action_search);
            if (itemSearch != null) {
                onQueryTextListenerSearch = new SearchView.OnQueryTextListener() {

                    @Override
                    public boolean onQueryTextSubmit(String query) {
                        return false;
                    }

                    @Override
                    public boolean onQueryTextChange(String newText) {
                        List<Song> songs = activityMain.getAllSongs();
                        List<Song> sifted = new ArrayList<>();
                        if (!newText.equals("")) {
                            for (Song song : songs) {
                                if (song.title.toLowerCase().contains(newText.toLowerCase())) {
                                    sifted.add(song);
                                }
                            }
                            recyclerViewAdapter.updateList(sifted);
                        } else {
                            recyclerViewAdapter.updateList(songs);
                        }
                        return true;
                    }
                };
                SearchView searchView = (SearchView) itemSearch.getActionView();
                searchView.setOnQueryTextListener(onQueryTextListenerSearch);
            }
        }
    }

    private void setUpRecyclerView() {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        recyclerViewSongList = binding.recyclerViewSongList;
        recyclerViewSongList.setLayoutManager(new LinearLayoutManager(recyclerViewSongList.getContext()));
        for (Song song : viewModelUserPickedSongs.getUserPickedSongs()) {
            song.setSelected(true);
        }
        recyclerViewAdapter = new RecyclerViewAdapterSelectSongs(
                this, activityMain.getAllSongs());
        recyclerViewSongList.setAdapter(recyclerViewAdapter);
    }

    @Override
    public void onResume() {
        super.onResume();
        setUpToolbar();
        updateFAB();
    }

    private void updateFAB() {
        viewModelActivityMain.setFabImage(R.drawable.ic_check_white_24dp);
        viewModelActivityMain.setFABText(R.string.fab_done);
        viewModelActivityMain.showFab(true);
        onClickListenerFABFragmentSelectSongs = (view) -> {
            NavController navController = NavHostFragment.findNavController(this);
            if (navController.getCurrentDestination().getId() == R.id.FragmentSelectSongs) {
                navController.popBackStack();
            }
        };
        viewModelActivityMain.setFabOnClickListener(onClickListenerFABFragmentSelectSongs);
    }

    @Override
    public List<Song> getUserPickedSongs() {
        return viewModelUserPickedSongs.getUserPickedSongs();
    }

    @Override
    public void removeUserPickedSong(Song song) {
        viewModelUserPickedSongs.removeUserPickedSong(song);
    }

    @Override
    public void addUserPickedSong(Song song) {
        viewModelUserPickedSongs.addUserPickedSong(song);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ActivityMain activityMain = (ActivityMain) requireActivity();
        activityMain.unregisterReceiver(broadcastReceiver);
        broadcastReceiver = null;
        viewModelActivityMain.setFabOnClickListener(null);
        onClickListenerFABFragmentSelectSongs = null;
        recyclerViewSongList.setAdapter(null);
        recyclerViewAdapter = null;
        recyclerViewSongList = null;
        Toolbar toolbar = activityMain.findViewById(R.id.toolbar);
        Menu menu = toolbar.getMenu();
        if (menu != null) {
            MenuItem itemSearch = menu.findItem(R.id.action_search);
            SearchView searchView = (SearchView) itemSearch.getActionView();
            searchView.setOnQueryTextListener(null);
            searchView.onActionViewCollapsed();
        }
        onQueryTextListenerSearch = null;
        viewModelUserPickedSongs = null;
        viewModelActivityMain = null;
        binding = null;
    }

}

File: app/src/main/java/com/fourthFinger/pinkyPlayer/fragments/FragmentSelectSongs.kt
package com.fourthFinger.pinkyPlayer.fragments

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.fourthFinger.pinkyPlayer.KeyboardUtil
import com.fourthFinger.pinkyPlayer.R
import com.fourthFinger.pinkyPlayer.activity_main.ActivityMain
import com.fourthFinger.pinkyPlayer.activity_main.ViewModelActivityMain
import com.fourthFinger.pinkyPlayer.databinding.RecyclerViewSongListBinding
import com.fourthFinger.pinkyPlayer.random_playlist.Song

class FragmentSelectSongs : Fragment(), RecyclerViewAdapterSelectSongs.ListenerCallbackSelectSongs {

    private var _binding: RecyclerViewSongListBinding? = null
    private val binding get() = _binding!!

    private val viewModelActivityMain by activityViewModels<ViewModelActivityMain>()
    private val viewModelPlaylists by activityViewModels<ViewModelPlaylists>()

    private var recyclerViewSongList: RecyclerView? = null
    private lateinit var recyclerViewAdapter: RecyclerViewAdapterSelectSongs

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        _binding = RecyclerViewSongListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        KeyboardUtil.hideKeyboard(view)
        viewModelActivityMain.setActionBarTitle(resources.getString(R.string.select_songs))
        setUpRecyclerView()
    }

    private fun setUpRecyclerView() {
        recyclerViewSongList = binding.recyclerViewSongList
        recyclerViewSongList?.layoutManager = LinearLayoutManager(requireContext())
        // TODO respond to LiveData of all songs
        recyclerViewAdapter = viewModelPlaylists.getAllSongs()?.let {
            RecyclerViewAdapterSelectSongs(this, it)
        }?: RecyclerViewAdapterSelectSongs(this, listOf())
        recyclerViewSongList?.adapter = recyclerViewAdapter
    }

    override fun onStart() {
        super.onStart()
        setUpBroadcastReceiver()
    }

    private fun setUpBroadcastReceiver() {
        val intentFilter = IntentFilter()
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT)
        intentFilter.addAction(requireActivity().resources.getString(
                R.string.broadcast_receiver_action_service_connected))
        requireActivity().registerReceiver(broadcastReceiver, intentFilter)
    }

    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action: String? = intent.action
            if (action != null) {
                if (action == resources.getString(
                        R.string.broadcast_receiver_action_service_connected)) {
                    setUpRecyclerView()
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        updateFAB()
    }

    private fun updateFAB() {
        viewModelActivityMain.setFabImage(R.drawable.ic_check_white_24dp)
        viewModelActivityMain.setFABText(R.string.fab_done)
        viewModelActivityMain.showFab(true)
        viewModelActivityMain.setFabOnClickListener{
            val navController: NavController = NavHostFragment.findNavController(this)
            if (navController.currentDestination?.id == R.id.FragmentSelectSongs) {
                navController.popBackStack()
            }
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.getItem(ActivityMain.MENU_ACTION_SEARCH_INDEX).isVisible = true
        val itemSearch = menu.findItem(R.id.action_search)
        if (itemSearch != null) {
            val onQueryTextListenerSearch = object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String?): Boolean {
                    return false
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    filterSongs(newText)
                    return true
                }
            }
            val searchView = itemSearch.actionView as SearchView
            searchView.setOnQueryTextListener(onQueryTextListenerSearch)
        }
    }

    private fun filterSongs(newText: String) {
        if (newText.isNotEmpty()) {
            val sifted = viewModelPlaylists.filterAllSongs(newText)
            recyclerViewAdapter.updateList(sifted)
        } else {
            viewModelPlaylists.getUserPickedPlaylist()?.getSongs()?.let {
                recyclerViewAdapter.updateList(it)
            }
        }
    }

    // TODO should be able to remove
    override fun getUserPickedSongs(): List<Song> {
        return viewModelPlaylists.getUserPickedSongs()
    }

    override fun songUnselected(song: Song) {
        viewModelPlaylists.songUnselected(song)
    }

    override fun songSelected(song: Song) {
        viewModelPlaylists.songSelected(song)
    }

    override fun onStop() {
        super.onStop()
        requireActivity().unregisterReceiver(broadcastReceiver)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModelActivityMain.setFabOnClickListener(null)
        val toolbar: Toolbar = requireActivity().findViewById(R.id.toolbar)
        val menu = toolbar.menu
        if (menu != null) {
            val itemSearch = menu.findItem(R.id.action_search)
            val searchView = itemSearch.actionView as SearchView
            searchView.setOnQueryTextListener(null)
            searchView.onActionViewCollapsed()
        }
    }

}